staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,247 @@ limitations under the License.
 	"time"
 
 	restful "github.com/emicklei/go-restful/v3"
 
-	"k8s.io/klog/v2"
-
 	"k8s.io/apiserver/pkg/server"
+	"k8s.io/klog/v2"
 	v1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
 	"k8s.io/kube-openapi/pkg/aggregator"
 	"k8s.io/kube-openapi/pkg/builder"
+	"k8s.io/kube-openapi/pkg/cached"
 	"k8s.io/kube-openapi/pkg/common"
 	"k8s.io/kube-openapi/pkg/common/restfuladapter"
 	"k8s.io/kube-openapi/pkg/handler"
 	"k8s.io/kube-openapi/pkg/validation/spec"
 )
 
+var ErrNotFound = errors.New("resource not found")
+
 // SpecAggregator calls out to http handlers of APIServices and merges specs. It keeps state of the last
 // known specs including the http etag.
 type SpecAggregator interface {
-	AddUpdateAPIService(handler http.Handler, apiService *v1.APIService) error
-	UpdateAPIServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error
-	RemoveAPIServiceSpec(apiServiceName string) error
-	GetAPIServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool)
-	GetAPIServiceNames() []string
+	AddAPIService(apiService *v1.APIService, handler http.Handler) error
+	UpdateAPIService(apiServiceName string) error
+	RemoveAPIService(apiServiceName string) error
 }
 
 const (
 	aggregatorUser                = "system:aggregator"
-	specDownloadTimeout           = 60 * time.Second
-	localDelegateChainNamePrefix  = "k8s_internal_local_delegation_chain_"
-	localDelegateChainNamePattern = localDelegateChainNamePrefix + "%010d"
+	specDownloadTimeout           = time.Minute
+	localDelegateChainNamePattern = "k8s_internal_local_delegation_chain_%010d"
 
 	// A randomly generated UUID to differentiate local and remote eTags.
 	locallyGeneratedEtagPrefix = "\"6E8F849B434D4B98A569B9D7718876E9-"
 )
 
-// IsLocalAPIService returns true for local specs from delegates.
-func IsLocalAPIService(apiServiceName string) bool {
-	return strings.HasPrefix(apiServiceName, localDelegateChainNamePrefix)
+// openAPISpecInfo is used to store OpenAPI spec with its priority.
+// It can be used to sort specs with their priorities.
+type openAPISpecInfo struct {
+	apiService v1.APIService
+	// This is the cache used to hold the specs fresh from the apiservices
+	spec cached.Replaceable[*spec.Swagger]
+
+	// The downloader is used only for non-local apiservices to
+	// re-update the spec every so often.
+	downloader cached.Data[*spec.Swagger]
 }
 
-// GetAPIServiceNames returns the names of APIServices recorded in specAggregator.openAPISpecs.
-// We use this function to pass the names of local APIServices to the controller in this package,
-// so that the controller can periodically sync the OpenAPI spec from delegation API servers.
-func (s *specAggregator) GetAPIServiceNames() []string {
-	names := make([]string, 0, len(s.openAPISpecs))
-	for key := range s.openAPISpecs {
-		names = append(names, key)
+type specAggregator struct {
+	mutex sync.Mutex
+
+	// Map of API Services' OpenAPI specs by their name
+	openAPISpecs map[string]*openAPISpecInfo
```

Comment:
```suggestion
	specsByAPIServiceName map[string]*openAPISpecInfo
```

or just specs. The `openAPI` prefixes don't add anything in this file.